script-runner/TestCases.hs

```diff
@@ -50,6 +60,68 @@ data ExpectedResult = SuccessFullUpdate | FailedProposalUpdate deriving (Show, E
 logMsg :: String -> PocMode ()
 logMsg = liftIO . hPutStrLn stderr
 
+{-# ANN testHardForkMechanism ("HLint: ignore Reduce duplication" :: Text) #-}
+testHardForkMechanism :: Script ()
+testHardForkMechanism = do
+  genesisConfig <- getGenesisConfig
+  let
+    proposal :: Dict HasConfigurations -> Diffusion PocMode -> PocMode ()
+    proposal Dict diffusion = do
+      let
+        keyIndex :: Int
+        keyIndex = 0
+        blockVersion = BlockVersion 0 1 0
+        softwareVersion = SoftwareVersion (ApplicationName "cardano-sl") 1
+        blockVersionModifier :: BlockVersionModifier
+        blockVersionModifier = def { bvmUnlockStakeEpoch = Just obftEraFlagValue }
+      doUpdate diffusion genesisConfig keyIndex blockVersion softwareVersion blockVersionModifier
+    expectedResult = SuccessFullUpdate
+  onStartup $ \Dict _diffusion -> do
+    stateDir <- view acStatePath
+    loadNKeys stateDir 4
+  on (1,2) proposal
+  on (2,6) $ \Dict _diffusion -> do
+    uc <- view (lensOf @UpdateConfiguration)
+    proposals <- GS.getConfirmedProposals uc Nothing
+    case (proposals, expectedResult) of
+      ([], FailedProposalUpdate) -> do
+        logMsg "expected failed proposal, passing test"
+        endScript ExitSuccess
+      (_, FailedProposalUpdate) -> do
+        logMsg "expected failure, but proposal was accepted!"
+        endScript $ ExitFailure 1
+      ([], _) -> do
+        logMsg "expected proposal to pass, but it didnt"
+        endScript $ ExitFailure 2
+      ([_one], SuccessFullUpdate) -> do
+        stateDir <- view acStatePath
+        opts <- view acScriptOptions
+        let
+          -- the config for the script-runner is mirrored to the nodes it starts
+          cfg = opts ^. srCommonNodeArgs . CLI.commonArgs_L
+        newConfiguration <- liftIO $ mutateConfigurationYaml (cfg ^. CLI.configurationOptions_L . cfoFilePath_L) (cfg ^. CLI.configurationOptions_L . cfoKey_L) mutateConfigurationForObft
+        liftIO $ BS.writeFile (T.unpack $ stateDir <> "/configuration2.yaml") newConfiguration
+        let
+          cfg2 = cfg & CLI.configurationOptions_L . cfoFilePath_L .~ (T.unpack $ stateDir <> "/configuration2.yaml")
+        forAllNodes_ $ \node -> do
+          stopNodeByName (Core, node)
+          startNode $ NodeInfo node Core stateDir (stateDir <> "/topology.yaml") cfg2
+      (_toomany, SuccessFullUpdate) -> do
+        logMsg "expected 1 proposal to pass, but >1 have passed"
+        endScript $ ExitFailure 3
+  on (3,10) $ \Dict _diffusion -> do
+    bvd <- gsAdoptedBVData
+    case (bvdUnlockStakeEpoch bvd == obftEraFlagValue) of
+      True -> do
+        liftIO $ hPutStrLn stderr "test passed"
```

Comment:
Do we have to write this to stderr for some reason? This might be a message we'd still like to see even when suppressing stderr output, for example (`2> /dev/null`).

Reply:
Good call, this is a holdover lifted from [here](https://github.com/input-output-hk/cardano-sl/pull/4061/files#diff-d9508ff47fca2564466ff5541ffe806bL105) which should also change.

Unless @cleverca22 has a preference for `stderr` - [`logMsg`](https://github.com/input-output-hk/cardano-sl/pull/4061/files#diff-d9508ff47fca2564466ff5541ffe806bR61) seems to express that preference.

Reply:
@intricate and I decided to defer this to a later time, if necessary, so we can keep moving on OBFT.